compiler/rustc_mir_build/src/build/expr/into.rs

```diff
@@ -344,25 +344,63 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
-                            }
-                        })
-                        .collect()
-                } else {
-                    field_names.filter_map(|n| fields_map.get(&n).cloned()).collect()
+                let variant = adt_def.variant(variant_index);
+                let field_names = variant.fields.indices();
+
+                let fields = match base {
+                    AdtExprBase::None => {
+                        field_names.filter_map(|n| fields_map.get(&n).cloned()).collect()
+                    }
+                    AdtExprBase::Base(FruInfo { base, field_types }) => {
+                        let place_builder = unpack!(block = this.as_place_builder(block, *base));
+
+                        // MIR does not natively support FRU, so for each
+                        // base-supplied field, generate an operand that
+                        // reads it from the base.
+                        iter::zip(field_names, &**field_types)
+                            .map(|(n, ty)| match fields_map.get(&n) {
+                                Some(v) => v.clone(),
+                                None => {
+                                    let place =
+                                        place_builder.clone_project(PlaceElem::Field(n, *ty));
+                                    this.consume_by_copy_or_move(place.to_place(this))
+                                }
+                            })
+                            .collect()
+                    }
+                    AdtExprBase::DefaultFields(field_types) => {
+                        iter::zip(field_names, &**field_types)
+                            .map(|(n, ty)| match fields_map.get(&n) {
+                                Some(v) => v.clone(),
+                                None => match variant.fields[n].value {
+                                    Some(def) => {
+                                        let value =
+                                            this.tcx.instantiate_and_normalize_erasing_regions(
+                                                args,
+                                                this.infcx.typing_env(this.param_env),
+                                                Const::from_unevaluated(this.tcx, def),
+                                            );
+                                        let value = match value.eval(
+                                            this.tcx,
+                                            this.infcx.typing_env(this.param_env),
+                                            expr_span,
+                                        ) {
+                                            Ok(val) => Const::Val(val, value.ty()),
+                                            Err(_) => value,
+                                        };
+                                        this.literal_operand(expr_span, value)
+                                    }
```

Comment:
This is the meat and potatos of the feature: if the code is `Struct { .. }`, when go through every field with defaults `const`s and evaluate them. This is where I most suspect I might be missing some handling we should have.